pax-exam-container-carbon/org.wso2.carbon.container/src/main/java/org/wso2/carbon/container/ArchiveExtractor.java

```diff
@@ -0,0 +1,163 @@
+        if (sourceURL.getProtocol().equals("file")) {
+            String file = sourceURL.getFile();
+            if (file.endsWith(".zip")) {
+                extractZipDistribution(sourceURL, targetDirectory);
+            } else if (file.endsWith(".tar.gz")) {
+                extractTarGzDistribution(sourceURL, targetDirectory);
+            } else {
+                throw new TestContainerException(
+                        "Unknown packaging of distribution; only zip or tar.gz could be handled.");
+            }
+            return;
+        }
+        if (sourceURL.toExternalForm().endsWith("/zip")) {
+            extractZipDistribution(sourceURL, targetDirectory);
+        } else if (sourceURL.toExternalForm().endsWith("/tar.gz")) {
+            extractTarGzDistribution(sourceURL, targetDirectory);
+        } else {
+            throw new TestContainerException(
+                    "Unknown packaging; only zip or tar.gz could be handled. URL was " + sourceURL);
+        }
+    }
+
+    /**
+     * Extract zip file in the system to a target Directory.
+     *
+     * @param path            path of the archive to be extract
+     * @param targetDirectory where to extract to
+     * @throws IOException on I/O error
+     */
+    public static void extract(Path path, File targetDirectory) throws IOException {
+        if (path.toString().endsWith(".zip")) {
+            extract(new ZipArchiveInputStream(new FileInputStream(path.toFile())), targetDirectory);
+        } else {
+            throw new TestContainerException("Unknown packaging of distribution; only zip can be handled.");
+        }
+    }
+
+    private static void extractTarGzDistribution(URL sourceDistribution, File targetDirectory) throws IOException {
+        File uncompressedFile = File.createTempFile("uncompressedTarGz-", ".tar");
+        extractGzArchive(sourceDistribution.openStream(), uncompressedFile);
+        extract(new TarArchiveInputStream(new FileInputStream(uncompressedFile)), targetDirectory);
+        FileUtils.forceDelete(uncompressedFile);
+    }
+
+    private static void extractZipDistribution(URL sourceDistribution, File targetDirectory) throws IOException {
+        extract(new ZipArchiveInputStream(sourceDistribution.openStream()), targetDirectory);
+    }
+
+    private static void extractGzArchive(InputStream tarGz, File tar) throws IOException {
+        BufferedInputStream in = new BufferedInputStream(tarGz);
+
+        try (
+                FileOutputStream out = new FileOutputStream(tar);
+                GzipCompressorInputStream gzIn = new GzipCompressorInputStream(in)
+        ) {
+            final byte[] buffer = new byte[1000];
+            int n;
+            while (-1 != (n = gzIn.read(buffer))) {
+                out.write(buffer, 0, n);
+            }
+        }
+    }
+
+    private static void extract(ArchiveInputStream is, File targetDir) throws IOException {
+        try {
```

Comment:
It seems this usage is not a good practice, Creating a stream from one method and closing it in another.
Can you try to pass the File instead, In this method, you can create the stream and close. (use autoclosable feature)